x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/highlight/SemanticTextHighlighter.java

```diff
@@ -0,0 +1,208 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link Highlighter} designed for the {@link SemanticTextFieldMapper}.
+ * This highlighter extracts semantic queries and evaluates them against each chunk produced by the semantic text field.
+ * It returns the top-scoring chunks as snippets, optionally sorted by their scores.
+ */
+public class SemanticTextHighlighter implements Highlighter {
+    public static final String NAME = "semantic";
+
+    private record OffsetAndScore(int offset, float score) {}
+
+    @Override
+    public boolean canHighlight(MappedFieldType fieldType) {
+        if (fieldType instanceof SemanticTextFieldMapper.SemanticTextFieldType) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public HighlightField highlight(FieldHighlightContext fieldContext) throws IOException {
+        SemanticTextFieldMapper.SemanticTextFieldType fieldType = (SemanticTextFieldMapper.SemanticTextFieldType) fieldContext.fieldType;
+        if (fieldType.getEmbeddingsField() == null) {
+            // nothing indexed yet
+            return null;
+        }
+
+        final List<Query> queries = switch (fieldType.getModelSettings().taskType()) {
+            case SPARSE_EMBEDDING -> extractSparseVectorQueries(
+                (SparseVectorFieldType) fieldType.getEmbeddingsField().fieldType(),
+                fieldContext.query
+            );
+            case TEXT_EMBEDDING -> extractDenseVectorQueries(
+                (DenseVectorFieldType) fieldType.getEmbeddingsField().fieldType(),
+                fieldContext.query
+            );
+            default -> throw new IllegalStateException(
+                "Wrong task type for a semantic text field, got [" + fieldType.getModelSettings().taskType().name() + "]"
+            );
+        };
+        if (queries.isEmpty()) {
+            // nothing to highlight
+            return null;
+        }
+
+        int numberOfFragments = fieldContext.field.fieldOptions().numberOfFragments() <= 0
+            ? 1 // we return the best fragment by default
+            : fieldContext.field.fieldOptions().numberOfFragments();
+
+        List<OffsetAndScore> chunks = extractOffsetAndScores(
+            fieldContext.context.getSearchExecutionContext(),
+            fieldContext.hitContext.reader(),
+            fieldType,
+            fieldContext.hitContext.docId(),
+            queries
+        );
+        if (chunks.size() == 0) {
+            return null;
+        }
+
+        chunks.sort(Comparator.comparingDouble(OffsetAndScore::score).reversed());
+        int size = Math.min(chunks.size(), numberOfFragments);
+        if (fieldContext.field.fieldOptions().scoreOrdered() == false) {
+            chunks = chunks.subList(0, size);
+            chunks.sort(Comparator.comparingInt(c -> c.offset));
+        }
```

Comment:
Could we refactor this so that we only sort once?

Reply:
We can't do that. ;)
The first step is to get the best snippets, then we prune them, and finally, we restore the passages to their original order.
One optimization here could be to use a priority queue in the first step, but since the number of chunks should be small, I didn't bother with  that.